core/meta/object-descriptor.js

```diff
@@ -357,35 +358,121 @@ var ObjectDescriptor = exports.ObjectDescriptor = Montage.specialize( /** @lends
+                    if (descriptor.owner === this) {
+                        this._propertyDescriptors.push(descriptor);
+                        this._propertyDescriptorsTable.set(descriptor.name,  descriptor);
+                    } else if (!this._propertyDescriptorsTable.has(descriptor.name)) {
+                        this._propertyDescriptors.push(descriptor);
+                        this._propertyDescriptorsTable.set(descriptor.name,  descriptor);
+                    }
+                    descriptor._owner = descriptor._owner || this;
+                }
             }
+        }
+    },
 
+    _preparePropertyDescriptorsCache: {
+        value: function () {
+            var ownDescriptors = this._ownPropertyDescriptors,
+                isReady = true,
+                descriptor, i, n;
+            
+            if (!this._propertyDescriptors)  {
+                
+                for (i = 0, n = ownDescriptors.length; i < n && isReady; ++i) {
+                    descriptor = ownDescriptors[i];
+                    isReady = !!(descriptor && descriptor.name);
+                }
+    
+                if (isReady) {
+                    this._propertyDescriptors = [];
+                    this._propertyDescriptorsTable.clear();
+                    for (i = 0, n = ownDescriptors.length; i < n && isReady; ++i) {
+                        descriptor = ownDescriptors[i];
+                        descriptor._owner = this;
+                        this._propertyDescriptors.push(descriptor);
+                        this._propertyDescriptorsTable.set(descriptor.name,  descriptor);
+                    }
+                    this.addRangeAtPathChangeListener("_ownPropertyDescriptors", this, "_handlePropertyDescriptorsRangeChange");
+                    this.addRangeAtPathChangeListener("parent.propertyDescriptors", this, "_handlePropertyDescriptorsRangeChange");
+                }
+            }
         }
     },
 
     /**
-     * @returns {Array.<PropertyDescriptor>}
+     * PropertyDescriptors for this object descriptor, not including those 
+     * provided by this.parent
+     *
+     * @property {Array<PropertyDescriptor>} value
+     */
+    _ownPropertyDescriptors: {
+        value: null
+    },
+
+    _propertyDescriptors: {
+        value: null
+    },
+
+    /**
+     * PropertyDescriptors associated to this object descriptor, including those 
+     * provided by this.parent.
+     *
+     * @property {Array<PropertyDescriptor>} value
      */
     propertyDescriptors: {
         get: function () {
-            var propertyDescriptors = [];
-            propertyDescriptors = propertyDescriptors.concat(this._propertyDescriptors);
-            if (this.parent) {
-                propertyDescriptors = propertyDescriptors.concat(this.parent.propertyDescriptors);
+            this._preparePropertyDescriptorsCache();
+            return this._propertyDescriptors || [];
```

Comment:
sounds funny to return an empty array every time and not assign it to this._propertyDescriptors the first time if this._propertyDescriptors isn't  a thing yet ?

Reply:
Agreed. It serves a purpose though so we'll need a new strategy for that. 

The issue is that we can't build out the propertyDescriptor caches until every propertyDescriptor has a name. You can see this at work in `_preparePropertyDescriptorsCache`

https://github.com/montagejs/montage/blob/master/core/meta/object-descriptor.js#L408-L435

At the time I avoided using a special flag to gate whether we should try to build the cache instead of using `this._propertyDescriptors` like I do here: 


```javascript
            if (!this._propertyDescriptors)  {
                
                for (i = 0, n = ownDescriptors.length; i < n && isReady; ++i) {
                    descriptor = ownDescriptors[i];
                    isReady = !!(descriptor && descriptor.name);
                }
```

Though on further reflection it does seem wasteful to repeatedly create a new array only in order to avoid one boolean. So we could instead have a flag such as `_propertyDescriptorsAreCached` or `_propertyDescriptorsAreReady`. Then that would change `_preparePropertyDescriptorsCache` to the following: 

```javascript
_preparePropertyDescriptorsCache: {
    value: function () {
        var ownDescriptors = this._ownPropertyDescriptors,
            isReady = true,
            descriptor, i, n;
        
        if (!this._propertyDescriptorsAreCached)  {
            
            for (i = 0, n = ownDescriptors.length; i < n && isReady; ++i) {
                descriptor = ownDescriptors[i];
                isReady = !!(descriptor && descriptor.name);
            }

            if (isReady) {
                this._propertyDescriptorsAreCached = true;
                this._propertyDescriptorsTable.clear();
                for (i = 0, n = ownDescriptors.length; i < n && isReady; ++i) {
                    descriptor = ownDescriptors[i];
                    descriptor._owner = this;
                    this._propertyDescriptors.push(descriptor);
                    this._propertyDescriptorsTable.set(descriptor.name,  descriptor);
                }
                this.addRangeAtPathChangeListener("_ownPropertyDescriptors", this, "_handlePropertyDescriptorsRangeChange");
                this.addRangeAtPathChangeListener("parent.propertyDescriptors", this, "_handlePropertyDescriptorsRangeChange");
            }
        }
    }
},

```

And allow the propertyDescriptors getter to be: 
```javascript
propertyDescriptors: {
        get: function () {
            if (!this._propertyDescriptors) {
                this._propertyDescriptors = [];
            }
            this._preparePropertyDescriptorsCache();
            return this._propertyDescriptors;
        }
    },
```


What do you think?